components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingChallengeQuestionsHandler.java

```diff
@@ -0,0 +1,411 @@
+                if (question.getQuestion().equals(challengeQuestion)) {
+                    UserChallengeAnswer questionAndAnswer = new UserChallengeAnswer();
+                    questionAndAnswer.setQuestion(question);
+                    questionAndAnswer.setAnswer(answersMap.get(s));
+                    questionsAndAnswers.add(questionAndAnswer);
+                }
+            }
+        }
+        return questionsAndAnswers.toArray(new UserChallengeAnswer[questionsAndAnswers.size()]);
+    }
+
+    /**
+     * Returns URL encoded String with challenge questions of the given user
+     *
+     * @param user Authenticated User.
+     * @return UTF-8 encoded URL with challenge questions.
+     */
+    private String getUrlEncodedChallengeQuestionsString(AuthenticatedUser user) {
+        StringBuilder challengeQuestionData = new StringBuilder();
+        List<ChallengeQuestion> challengeQuestionList = getChallengeQuestions(user);
+
+        if (challengeQuestionList == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Challenge Questions for found for tenant domain of the user: " + user.getUserName());
+            }
+            return null;
+        } else {
+            for (ChallengeQuestion question : challengeQuestionList) {
+                String setId = question.getQuestionSetId();
+                String questionId = question.getQuestionId();
+                String questionString = question.getQuestion();
+                String questionLocale = question.getLocale();
+                challengeQuestionData.append(setId).append("|").append(questionId).append("|").append
+                        (questionString).append("|").append(questionLocale).append("&");
+            }
+        }
+        String encodedData = null;
+        try {
+            encodedData = java.net.URLEncoder.encode(challengeQuestionData.toString(), "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            log.error("Error occurred while encoding Challenge question data as URL", e);
+        }
+        return encodedData;
+    }
+
+    /**
+     * This handles the PostAuthentication Requests for PostAuthnMissingChallengeQuestionHandler
+     *
+     * @param httpServletResponse   HTTP Servlet Response.
+     * @param authenticationContext Authentication Context
+     * @param user                  Authenticated User
+     * @return PostAuthnHandlerFlowStatus Flow status of the PostAuthentication Handler
+     */
+    private PostAuthnHandlerFlowStatus handlePostAuthenticationForMissingChallengeQuestionRequest(
+            HttpServletResponse httpServletResponse, AuthenticationContext authenticationContext,
+            AuthenticatedUser user) {
+
+        // If challenge questions are not requested redirect user to add challenge questions jsp page
+        String encodedData = getUrlEncodedChallengeQuestionsString(user);
+
+        if (encodedData == null || "".equals(encodedData)) {
```

Comment:
Rather than sending null for error scenarios in getChallengeQuestions and getUrlEncodedChallengeQuestionsString methods, it would be better to throw the error from those methods and handle it here.

Reply:
resolved